framework/src/bcs/NonlocalIntegratedBC.C

```diff
@@ -0,0 +1,151 @@
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "NonlocalIntegratedBC.h"
+#include "Assembly.h"
+#include "MooseVariable.h"
+#include "Problem.h"
+#include "SubProblem.h"
+#include "SystemBase.h"
+#include "MooseMesh.h"
+
+// libmesh includes
+#include "libmesh/threads.h"
+#include "libmesh/quadrature.h"
+
+template<>
+InputParameters validParams<NonlocalIntegratedBC>()
+{
+  InputParameters params = validParams<IntegratedBC>();
+  return params;
+}
+
+NonlocalIntegratedBC::NonlocalIntegratedBC(const InputParameters & parameters) :
+    IntegratedBC(parameters)
+{
+  _mesh.errorIfDistributedMesh("NonlocalIntegratedBC");
+  mooseWarning("NonlocalIntegratedBC is a computationally expensive experimental capability used only for integral terms.");
+}
+
+void
+NonlocalIntegratedBC::computeJacobian()
+{
+  DenseMatrix<Number> & ke = _assembly.jacobianBlock(_var.number(), _var.number());
+  _local_ke.resize(ke.m(), ke.n());
+  _local_ke.zero();
+
+  for (_j = 0; _j < _phi.size(); _j++) // looping order for _i & _j are reversed for performance improvement
+  {
+    getUserObjectJacobian(_var.number(), _var.dofIndices()[_j]);
+    for (_i = 0; _i < _test.size(); _i++)
+      for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+        _local_ke(_i, _j) += _JxW[_qp] * _coord[_qp] * computeQpJacobian();
+    }
+
+  ke += _local_ke;
+
+  if (_has_diag_save_in)
+  {
+    unsigned int rows = ke.m();
+    DenseVector<Number> diag(rows);
+    for (unsigned int i=0; i<rows; i++)
+      diag(i) = _local_ke(i,i);
+
+    Threads::spin_mutex::scoped_lock lock(Threads::spin_mtx);
+    for (const auto & var : _diag_save_in)
+      var->sys().solution().add_vector(diag, var->dofIndices());
+  }
+}
+
+void
+NonlocalIntegratedBC::computeJacobianBlock(unsigned int jvar)
+{
+  if (jvar == _var.number())
+    computeJacobian();
+  else
+  {
+    MooseVariable & jv = _sys.getVariable(_tid, jvar);
+    DenseMatrix<Number> & ke = _assembly.jacobianBlock(_var.number(), jvar);
+
+    for (_j = 0; _j < _phi.size(); _j++) // looping order for _i & _j are reversed for performance improvement
+    {
+      getUserObjectJacobian(jvar, jv.dofIndices()[_j]);
+      for (_i = 0; _i < _test.size(); _i++)
+        for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+          ke(_i, _j) += _JxW[_qp] * _coord[_qp] * computeQpOffDiagJacobian(jvar);
+    }
+  }
+}
+
+void
+NonlocalIntegratedBC::computeNonlocalJacobian()
+{
+  DenseMatrix<Number> & keg = _assembly.jacobianBlockNonlocal(_var.number(), _var.number());
+  // compiling set of global IDs for the local DOFs on the element
+  std::set<dof_id_type> local_dofindices(_var.dofIndices().begin(), _var.dofIndices().end());
```

Comment:
The use of a `std::set` here to obtain asymptotic _log n_ lookup might be slowing you down because of the very small size of the set. The overhead of the extra pointers and the red/black tree is a lot of machinery for small chunks of data. I do see that you are looking over a potentially large number of DOFs though below so it's not a terrible decision, it's just premature. A sorted vector might perform better.

I wouldn't worry about changing it for now. Generally you should profile before trying to perform optimization. We've been surprised many times before.